Add `dune init` command

shonfeder opened the pull request:
Towards completion of  #159

This implements creation of new projects, as discussed in https://github.com/ocaml/dune/issues/159#issuecomment-408607808 and using the more simplistic, hard-coded strings approach encouraged here https://github.com/ocaml/dune/issues/159#issuecomment-426724225

I think tests ensuring initialized projects can be build should be included in this PR, but I'd like to get the general shape of the implementation OKed before writing the test cases.

shonfeder committed:
Add `dune init` command

Signed-off-by: Shon Feder <shon.feder@gmail.com>
Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

shonfeder commented on src/init.ml:
```diff
@@ -0,0 +1,196 @@
+
+  let package_opam =
+    let name = Printf.sprintf "%s.opam" C.pname in
+    make_file name ""
```
I don't think there was any discussion in #1448 about generating the *content* of the `<project>.opam` file, tho I can definitely see the appeal. Would something like the following suffice (just taken from the opam documentation on [creating and publishing opam packages](https://opam.ocaml.org/doc/Packaging.html)?

```
opam-version: "2.0"
name: "project"
version: "0.1"
synopsis: "One-line description"
description: """
Longer description
"""
maintainer: "Name <email>"
authors: "Name <email>"
license: ""
homepage: ""
bug-reports: ""
dev-repo: ""
depends: [ ]
build: ["dune" "build" "-p" name]
```

rgrinberg replied:
Generating opam files should be done from the metadata of the dune-project and dune files. I suggest that we leave that to a different PR

shonfeder commented:
(Looks like I need to make several small changes for older OCaml versions. I'm looking in to that now.)

rgrinberg commented on src/init.ml:
```diff
@@ -0,0 +1,196 @@
+open! Stdune
+open! Import
+
+module Project = struct
```
Do we really need this extra `Project` wrapper? It doesn't seem like `Project.Kind.t` is any clearer than `Kind.t`. If anything it's a `Stanza.Kind.t`, but in which case `Stanza_kind.t` is just as good.`

rgrinberg commented on src/init.ml:
```diff
@@ -0,0 +1,196 @@
+end
+
+module Lib (C : InitContext) : ProjectDir = struct
+  let dir = Path.(append C.dir (of_string "src"))
```
Simpler is to use `Path.relative` here.

rgrinberg commented on src/init.ml:
```diff
@@ -0,0 +1,198 @@
+
+  let lib_ml =
+    let name = "lib.ml" in
+    let content = {|let hello name = Printf.sprintf "Hello, %s" name
```
An empty file seems better to me.

shonfeder replied:
I'll argue this point a bit, to be sure we consider the alternatives and their effect on the ecosystem. 

imo, the advantage of having a little placeholder code is that newcomers are thus given something in place to tweak and work with. E.g., with these place-holders, a brand new user can run

```
dune init my_project --bin
cd ./my_project
dune exec my_project
# Hello, project my_project
```

and have a working program that prints output in under a minutes. They can then tweak the code a bit to confirm things work as expected, and build from there.  The advantages of smoother on boarding and greater accessibility seem self-evident.

On the other hand, the cost to experienced users is negligible: they need only delete the lines they don't want. Moreover, once custom templates are implemented in subsequent additions, we can provide an easy override, so experienced user can generate blank files, or no files, or whatever. But it seems to we should set the default to be something useful for inexperienced users.

Finally, if we opt for the file to be empty, then why not just have no file at all?

rgrinberg replied:
>  But it seems to we should set the default to be something useful for inexperienced users.

I disagree. We should optimize the user experience for those have invoked the tool for more than a few times. Users do not stay beginners for long and these hints will quickly become obsolete.

> Finally, if we opt for the file to be empty, then why not just have no file at all?

That's probably best. Since we're creating a wrapped library, we can't really predicate the module name the user wants anyway. Perhaps we can have a module name to specify this?

For executables of course, we should keep the empty file.

I suppose our main disagreement comes from the philosophy of this new command. I expect this command to be useful in helping all users quickly write stanza boilerplate, rather than just a tool to help beginners. I'd be interested in coming up with a compromise that could satisfy both experienced and beginning users.

shonfeder replied:
I think that's a spot on articulation of the grounds for the disagreement. I think I had these two purposes both in mind, and was not keeping them adequately separated. On reflection, I agree we should separate out these concerns.

If you agree, I'll keep this PR and immediate future work focused on the general purpose utility. Once we have the tooling built out to reduce stanza boilerplate and facilitate composable (likely configurable) generation of common stanzas, we can use it to make a very nice, dedicated beginners tool. The latter can be under its own command, perhaps as `dune starter`, `dune wizard`, or something like that. Hows that sound?

rgrinberg commented on src/init.ml:
```diff
@@ -0,0 +1,198 @@
+end
+
+let make_project_dir_exn (module D : ProjectDir) =
+  if Path.exists D.dir then
```
Isn't it simpler to have `mkdir -p` behavior here? If it the directory is there - great, let's proceed.

rgrinberg commented on src/init.ml:
```diff
@@ -0,0 +1,198 @@
+    let content = Printf.sprintf "\
+%s
+(name %s)
+(version 0.1)
```
The dune-project is already generated by dune. In any case, the version number here isn't correct. I suggest that we skip (or fix) the generation of the dune-project file. Skipping seems preferable as it's already done by dune.

rgrinberg commented on src/init.ml:
```diff
@@ -0,0 +1,198 @@
+    Io.write_file ~binary:false path file.content
+end
+
+module type InitContext = sig
```
In dune, the coding convention is `Snake_case`. It's best that we stick to it.

rgrinberg reviewed (changes requested):
I don't quite understand the necessity of functor or first class modules here. It seems like plain old records would work just as well here.

rgrinberg commented on src/init.ml:
```diff
@@ -0,0 +1,198 @@
+    "@{<ok>Success@}: %s %s project initialized\n"
+    C.pname (Project.Kind.to_string C.kind)
+
+let print_hints (module C : InitContext) =
```
IMO, these hints are annoying for non new users. If the feature is aimed at experienced users as well, we should avoid them.

rgrinberg replied:
These particular hints are less annoying than the commented lines btw. As they don't require manual removal. If you'd prefer, you can keep these but remove the commented stuff.

The library and executable stanzas should already have the  key options exposed via `$ dune init --help` and down the road, once we have auto completion for dune files, then those commented lines will be even less useful.

shonfeder commented:
I opted for modules to avoid having to thread the name and path parameters through a bunch of functions. However, I can definitely understand that the end result seems quite heavy duty given the simplicity of the feature added! Moreover, if we're not generating any content for the files, there's less need to thread those parameters through. I'll re-structure for a more lightweight approach.

rgrinberg commented:
> I opted for modules to avoid having to thread the name and path parameters through a bunch of functions. However, I can definitely understand that the end result seems quite heavy duty given the simplicity of the feature added! Moreover, if we're not generating any content for the files, there's less need to thread those parameters through. I'll re-structure for a more lightweight approach.

Thanks, we have situations in dune where we have a little too much boilerplate and we go with functors but the situation here is nowhere near as acute. The only case where we use first class modules are plugins, so introducing it here to get rid of a boilerplate is certainly unprecedented. I think the code will look quite clean with records by the way so boilerplate won't even be an issue.

shonfeder commented:
Notes on a few of the changes:

- Replaced the baroque first class modules with records (I think I'm overeager for 1ML 😋).
- Replaced the check for a preexisting project directory with checks for preexisting files during their creation. Now, if, e.g., a file exists at `<project>/bin/main.ml`, it will report this fact and skip creation. This protects against overwriting files (as the check for the root directory was meant to), but also allow initializing existing projects by fleshing out missing dune files that may be needed.
- Removed the `<project>/src/lib.ml` file, but sneakily kept a "Hello, World" line in `<project>/bin/main.ml`. I think this innocuous: It doesn't introduce the same complications as the content in `<project>/src/lib.ml` but it still provides the benefit of being able to get a printing program with one command. However, I'll remove this line without complaint if @rgrinberg is opposed.

rgrinberg commented:
Okay, looks really nice. Good work @shonfeder. Before I merge I'm planning the following UI changes:

* Change `$ dune init foo --lib` to `$ dune init lib foo`. The reason for this is that we'd like `$ dune init lib --help` to only show the relevant options for `lib`.

* For the executable template, we don't want an associated library. I think a single dune + source file is enough.

* Let's remove the generation of all .opam files. First of all, it will be done properly from dune files later and second of all it's not helpful for non public libs and binaries anyway. And the new `@all` alias makes building private libraries and executables much easier so there's even less incentive for this to be the default.

* Add the ability to put paths in the argument `$ dune init <lib|exe> foo/bar` should create a `<lib|exe>` named `bar` inside `foo`. There's no need directory nesting past this I believe.

If you see any problems or propose any changes to these suggestions, please feel free to discuss.

If you'd like some help in implementing some of this stuff, I'll gladly help out as well as I realize that some of these change aren't small.

We'll also need to add a test suite for this feature. I suggest that we add this test suite before we start any of the changes I've proposed. If you need help here, I'm happy to point you in the right direction or put up a few tests.

emillon commented on src/init.ml:
```diff
@@ -0,0 +1,162 @@
+
+let make_project_dirs (c : Init_context.t) =
+  (* dirs that are always included in a project *)
+  let project_dirs =
```
What do you think about naming this `always_included_dirs`? that way you can drop the comment above and remove the shadowing below.

emillon commented on bin/main.ml:
```diff
@@ -809,6 +809,44 @@ let utop =
+    let doc = "The project is intended to be a library" in
+    (Kind.Library, Arg.info ["lib"; "library"] ~doc)
+
+  let bin =
```
I don't think that these will be the only types of projects, so I suggest refactoring that into `--kind lib` / `--kind bin`. This also removes the weird quirk about what happens when you pass `--lib --bin` (it seems that all are ignored except the last one). What do you think about this?

emillon replied:
I realize that this does not go in the same direction as what @rgrinberg suggested below - the point he makes about having specific docs & flags is a good one, not sure how to reconcile both.

rgrinberg replied:
Separate sub commands for every kind are preferable as they will allow us to have more specific `--help` for every sub command as most of the kinds have specific options.

emillon replied:
And they prevent passing several kinds so I'm 👍 for that.

shonfeder replied:
I do think nested sub-commands generally make for nicer UI, so I am on board for this change.

emillon commented on bin/main.ml:
```diff
@@ -809,6 +809,44 @@ let utop =
+             to the current working directory.|}
+      ; `P {|The flags $(b,--lib) and $(b,--bin) specify whether the created
+             project should be set up as a library or as an executable program
+             (respectively). The defaults behavior is to create libraries.|}
```
Typo: "the default behavior"

emillon reviewed (commented):
Thanks! This nex feature will probably open up a healthy amount of bikeshedding but this is probably necessary!

rgrinberg commented:
By the way, this issue is relevant to make the opam files obsolete: https://github.com/ocaml/dune/issues/1112

There are no immediate plans to implement this (we are looking for volunteers), but it shouldn't be too hard and hopefully will be present in the next version or two

shonfeder commented on src/init.ml:
```diff
@@ -0,0 +1,162 @@
+    (public_name %s)
+    ;; (inline_tests)
+    ;; (libraries core)
+    ;; (preprocess (pps ppx_inline_test))
```
Oh, my bad! I should have removed them as part of resolving https://github.com/ocaml/dune/pull/1448#discussion_r225009978. I agree that this kind of hand-holding is better left to tutorial material or to a special-purpose tutorial/wizard command.

shonfeder commented:
I have a question on one point, re: "Add the ability to put paths in the argument $ dune init <lib|exe> foo/bar". The current interface allows for `dune init --lib my_project path/to/some/location`. It seems to me preferable to keep this kind of behavior, rather than creating an idiosyncratic meaning for `/` and overloading the meaning of the first positional argument. So, it seems preferable to me to keep the current behavior.

Other than that, I don't have any problems with the proposed changes and I'm happy to make them. I will let know if I get stuck or bogged down, but I think I should be able to bring those changes about in the next several days.

Thanks for the guidance, corrections, and reviews so far, @rgrinberg and @emillon!

rgrinberg commented:
>  have a question on one point, re: "Add the ability to put paths in the argument $ dune init <lib|exe> foo/bar". The current interface allows for dune init --lib my_project path/to/some/location. It seems to me preferable to keep this kind of behavior, rather than creating an idiosyncratic meaning for / and overloading the meaning of the first positional argument. So, it seems preferable to me to keep the current behavior.

Well the current behavior isn't ideal as it's forcing a particular directory structure. Dune doesn't really recommend any directory structure and many existing projects and users will have their own preference regarding this. Therefore it seems best to make the feature usable with any kind of directory structure.

My suggestion was just one idea on how to add the flexibility to create w/e directory structure the user wants. I'm happy to hear other suggestions. The use cases I have in mind are:

* How do I create a library in the current directory?
* How do I create a binary in an existing directory?

> Other than that, I don't have any problems with the proposed changes and I'm happy to make them. I will let know if I get stuck or bogged down, but I think I should be able to bring those changes about in the next several days.

Thanks, we appreciate your work.

shonfeder commented:
Ah, I think I understand what you are after now: we're not talking about the root directory of the project but the directory in which the `dune` file declaring the library is created. I can see the use.

@rgrinberg, according to the plan sketched out at https://github.com/ocaml/dune/issues/159#issuecomment-414177211, the two features you mention

- How do I create a library in the current directory?
- How do I create a binary in an existing directory?

would be supplied by the `dune add` command. The way discussed there (and what I understand many people to have been requesting), `dune init` should prime more holistic project development, and provide a quick and easy jumping off place. Thus, I think it makes sense to have a second command dedicated to small utility operations like creating a dune file of a certain sort sort in such and such a directory.

rgrinberg commented:
It seems like `$ dune init` and `$ dune add` have a very overlapping feature set however. In fact, if `$ dune init` was idempotent, I can't think of any situation where I'd want to use `$ dune add`.

Also, I'm not very sure how `$ dune init` maps to a single project by the way. Most of the projects that I deal define more more than just 1 executables and/or a library. Having `$ dune init` be used once and `$ dune add` n times makes the latter the far more useful command to implement.

Here's another suggestion, if we want `$ dune init` to initialize a whole project skeleton for us, then we can have a `$ dune init project` subcommand. I agree that such a command should indeed create a directory.

By the way, I think it would help to have a type like:

```ocaml
type common =
  { name : string
  ; in_ : Path.t
  }
type public = bool
type inline_tests
type init = (* pick your own name here *)
  | Executable of common * public * libraries
  | Library of common * public * libraries * inline_tests
  | Test of common * Lib_name.t (* the library that we are testing *)
  | Project of common (* This constructor would include Library + Test for example *)
```

The above makes it very clear what arguments anything that we can `init` can accept. It's just a sketch though, I wouldn't necessarily use tuples for example.

shonfeder commented:
I am OK with ditching the `add` command and just putting that functionality in `init`. Rather than `dune init project lib ...` (pretty unwieldy) I suggest we follow the patterns of `cargo` and `stack` and use `init` for stuff that initializes things an existing project and use `new` for creating new projects. This will make it more concise, more accessible to new users, and more likely to be familiar to people coming from neighboring communities. (but this is all matter for another PR any how :) ).

I will build out the API for init along the lines you suggest: that makes good sense to me.

I'm hoping to have all the suggested changes in place and ready for review by the end of the weekend.

rgrinberg commented:
> Rather than dune init project lib ... (pretty unwieldy)

Sorry, my suggestion `$ due init {project,lib,exe}`. I did not suggest any kind of additional nesting. What I meant was that all the extra directories would only be created in the `project` init type.

I suppose the main difference between between our tool and cargo/stack is the once-only restriction of the templates that I'm not so fond of. It's true that a tool that quickly generates dune boilerplate vs a tool that starts a new project aren't the same thing. But it seems like at least a flexible `$ dune init` will handle the `new` with enough templates for it.

Btw, I have no experience with `cargo` but `$ cargo init` doesn't look very good. I've used stack a lot more and have found their custom templates quite useful, but their non composability is quite annoying. It's quite ugly to have hspec, servant, and servant-hspec templates for example. We can try and do better in this respect.

shonfeder commented:
I thought I recalled this being an issue.... according to dbuenzli/cmdliner#24, cmdliner doesn't support nested subcommands. Does anyone know of a reasonable way around this?

shonfeder commented:
Ah, in dbuenzli/cmdliner#95 there is mention of implementing sub commands manually using positional arguments. Before I dig in to that, I'd like to confirm that the increased complexity of manual implementation of subcommands here is worth the cost. We should perhaps also consider that, since `dune init` is no longer meant to create projects, it seems reasonable that someone might `dune init --lib --bin --test some_name .` to create test, executable, and library dune files with one command.... (or is the use of the single shared `some_name` for all a blocker?).

shonfeder commented:
Ah, ok. So we *do* want this to be a nested subcommand -- for the focused help message (and added flexibility, etc.). So, just to be sure we're on the same page and that I understand correctly what's wanted here: completion of this PR (in it's ideal form) is now dependent on implementing a custom extension to cmdliner's CLI parsing that supports nested subcommands with their own help menus. Correct?

rgrinberg commented:
Yup, that would be ideal. But the workaround you proposed is backwards compatible so I'm personally ok with delaying the fully featured CLI.

shonfeder commented:
OK, cool. In order to keep the scope of this PR smaller, I'll use the workaround (appropriately annotated with TODOs). I'll open a separate issue for the nested CLI feature.

shonfeder commented:
Sorry for the long radio silence. Work has been hectic.

I'm trying to get this to a state where I can wrap it up, but I think that the latest direction from @rgrinberg has expanded the scope somewhat, and I wanted to make my clear my intention on how to finish this PR off. 

If `dune init lib` is to support options like those sketched in the data type `Library of common * public * libraries * inline_tests` *and* be usefully idempotent, then I'll need to at least

1. Incorporate additional command line flags to specify options like `public`, `libraries`, and `inline_tests.
2. Load and parse and update existing dune files. (This is needed so that, e.g., running `dune init lib libfoo --libs foo bar --public` when `libfoo` already exists but without dependency `foo`, this can be added).

I'm part way along here, having sketched out the data structure and gotten part way with the CLI changes. I'm just digging in to the parsing part now. Overall, this should result in a more solid and flexible tool than the quick and easy start I began with.

As always, let me know if I've misunderstood something.

rgrinberg commented:
You've understood what has to be done, but I don't think all this stuff is necessary for a v1. I think only the 2nd point would be nice to have as it would be annoying to break for users later on.

shonfeder commented:
@rgrinberg I think I've pushed this quite close to v1 completion. Please see https://github.com/ocaml/dune/pull/1448/files#diff-f9ffe0462d56540693e478d557db4085 for the currently supported features and the test coverage of various edge cases.

The only outstanding feature I am aware of is the ability to update fields within an existing stanza (discussed in our last two comments).  This would allow the user to take an existing dune file like

```
(library
  (name foo)
  (libraries bar))
```

and run `dune init lib foo --libs baz --ppx ppx_foo` to update the file to read


```
(library
  (name foo)
  (libraries bar baz)
  (preprocess
    (ppx ppx_foo)))
```

afaict, this feature will require a fairly careful algorithm to support merging or updating dune stanzas. It also calls for some thought about it's semantics (e.g., should users be able to add/remove `inline_tests` and `public` fields this way? How does the CLI for that work?). 

While we could work that in to this PR, I think it risks making an already long PR much longer and prolonging the life of an already long-lived feature branch.

My proposal is to add a check for conflicting pre-existing stanzas, and error out with a "not yet supported" message if such a stanza is found. That would allow us to bring this PR to a close, and add the feature in a followup.

shonfeder commented:
I have added the proposed change to report to users that adding fields to existing stanzas is not yet supported. 

Please review at your convenience.

rgrinberg commented:
Thanks, will try to give this a full review today

rgrinberg commented on bin/init.ml:
```diff
@@ -0,0 +1,100 @@
+
+    let open Component in
+    let context = Init_context.make common_term.root path in
+    let common = Options.{name; libraries; pps; dir} in
```
I believe you just want `{Option. name; libraries, pps; dir}`. This will avoid triggering the unused open warning.

Another option that we use in dune is:

```
let common : Options.common = {name; libraries; pps; dir}
```

rgrinberg commented on src/dune_init.ml:
```diff
@@ -0,0 +1,382 @@
+    File.load_dune_file ~path:dir
+    |> File.add_stanzas stanza
+
+  let bin Options.{context; common; options : executable} =
```
Change this to:

```
let bin ({Options. context; common; options}) =
```

to restore compatibility with older OCaml.

rgrinberg commented:
Also, please lowercase all source file names. In particular, I'm refering to `Dune_init.mli`.

rgrinberg commented:
Some initial regarding usability:

* `$ dune init` is awesome. Great work.

* The extra `bin` directory created by `$ dune init exe foo` is annoying an unnecessary. This naming convention isn't universal anyway, let's just drop it. Same for the other commands

* The help appears to be wrong. When I do `$ dune init exe --help` I see irrelevant options such as `--inline-tests`. Those should not appear. Options such as `-w` should not appear either.

* When creating libraries, it should be possible to give a public name directly as well I believe. So instead of just `--public`, perhaps we should should also have `public_name` field.

shonfeder commented:
Thanks for the review, @rgrinberg! Working to address (and then do some additional cleanup) now.

re:

>  The help appears to be wrong. When I do `$ dune init exe --help` I see irrelevant options such as `--inline-tests`. Those should not appear. Options such as `-w` should not appear either.

There are two issues here:

1. The inclusion of irrelevant flags such as `--inline-tests` in `dune init exe --help` and `--public` in `dune init test --help` follows from the workaround needed in the absence of nested commands. I believe you OKed the workaround for the short term here https://github.com/ocaml/dune/pull/1448#issuecomment-431678614. As soon as this feature is complete, I hope to turn my attention to #1482, which would fix this defect.
    
    I can note which commands each option works with in the documentation for the flags, and I can add validation to check for use of mismatching options. Is this satisfactory for a v1?
2. The inclusion of `-w` is weird and surprising, but I think that falls outside the scope of this PR.

rgrinberg commented:
Indeed. It's been a while so I don't remember everything that was discussed. Yeah, that shouldn't be a blocker. Perhaps we should leave a message in those options descriptions regarding which subcommands they pertain too.

shonfeder commented:
(Updated my previous comment (which was posted to soon), to explain why I think the inclusion of the `-w` option seems to be outside of scope).

> Indeed. It's been a while so I don't remember everything that was discussed.

Totally understandable, and entirely my fault!

I will add the additional documentation and validation as discussed.

shonfeder commented:
Comments addressed.  Please take another look when you have time, @rgrinberg .

One final change I'd like to make: given the plan to support adding fields to existing components in the near future, I propose renaming this command to `add`. If there is already an existing `libfoo` and you want to add `libbar` as dependency `dune init lib libfoo --libs libbar` doesn't read well (what is being initialized?). But `dune add lib libfoo --libs libbar` can easily be read as "add libbar **to** the libraries of libfoo ". Are you OK with this change?

shonfeder commented:
I went ahead with the change from `dune init` to `dune add`. I can revert if there's strong objection, but I think there's a strong case for thinking the semantics work out better this way.

rgrinberg commented:
Let's get some responses regarding the usability of this feature from @emillon or @diml.

I slightly prefer `$ dune init` over `$ dune add` because `add` conveys really little meaning. Let's see if someone else comes up with a better name.

Khady commented:
I definitely prefer to have `init`. Because this command is mainly useful for people discovering dune and who want to setup a project quickly. In this context, the `init` command makes more sense, is easier to discover, and less scary. `add` could simply be an alias to `init`.

diml commented:
I prefer `dune init` as well. `add` feels too generic

shonfeder commented:
@Khady 

> Because this command is mainly useful for people discovering dune and who want to setup a project quickly. In this context, the init command makes more sense, is easier to discover, and less scary. add could simply be an alias to init.

I disagree. While the initial intent of this PR *was* to make just that kind of quick start utility, the way development was directed by @rgrinberg produced a utility that is significantly more low level (and flexible). I don't think new users will find the command useful: correct usage requires already possessing detailed knowledge of how to structure a project and what fields should be populated in a component. 

However, this utility does let us string together a couple commands (or a couple of underlying function invocations) to compose just a quick start utility! As soon as this PR is merged, I plan to follow up with a feature adding exactly that sort command (because that is what I was trying to acheive all along 😄). Now, we will we need a name for this quick start utility, and I was hoping to use `init` for that purpose (so `init` would be an alias to a chain of `add`s). 

My question: If we give the name `init` to *this* low level utility, what name would we like to see for the quick start utility?

@diml 

> `add` feels too generic

Doesn't this command do a very generic thing?  (Or, rather, *won't* it, once the additional feature discussed in https://github.com/ocaml/dune/pull/1448#issuecomment-467006261 is completed?) Currently, this command can be used to add a new dune file or add a component to an existing dune file. According to the plans for the immediate future, this command will also be used to add a library dependency, an `(inline_tests)` field, a `public_name` field, or a set of `ppx` preprocessors to an existing component. Given the tendency, it seems like we could easily include additional capabilities as well. I think the name should fit the functionality, and I am worried that `init` is *too specific*.

I should note, I am more opposed to naming this utility `init` than I am in favor of calling it `add`. I suspect that a primary use of this tool will be to add libraries and preprocessors (and likely additional qualifications in the future) to already defined components. So, before we have committed to calling this command `init`, let us consider carefully whether that name makes sense in this scenario:

- A user is working on a project with a `src/dune` file containing

    ```
    (library
     (name mylib)
     (public_name my_public_lib)
     (libraries core cmdliner async))
    ```
- The user decides they will make use of `notty`, so they execute (perhaps through an editor integration) `dune init mylib src --libs notty` in order to update `src/dune` to

    ```
    (library
     (name mylib)
     (public_name my_public_lib)
     (libraries core cmdliner async notty))
    ```

The dependency on the `notty` library has clearly been added to the `mylib` component, but what has been `init`ialized here? 

Can we think of any name that would make sense in this scenario but which you might like better than `add`?

rgrinberg commented:
@shonfeder you are free to add a template to `$ dune init` that would implement a bigger template out of the box. For example: `$ dune init project-lib-exe` that would initialize a full template out of the box. I just think that we need to have the proper low level commands in place first (exe, lib, test, etc.). It's not really necessary for this PR though.

diml commented:
One thing to consider is that we shouldn't add too many low-level toplevel commands. Otherwise it will hurt discoverability of commands and readability of man pages. CLI commands should be reserved to things the user is likely to type in the terminal.

We could group all advanced features under one toplevel command, for instance `dune expert edit-dune-file add ....`. However, this is not yet supported by cmdliner so this would feature would need to be added first. Alternatively, maybe low-level functionalities should be exposed as a library rather than via the CLI? For instance, editor support could be provided via a LSP server.

shonfeder commented:
@rgrinberg I'm confused. Nearly all of the work on this PR has been to implement the low level commands, and, I thought, we are now about 95% there. When you write "It's not really necessary for this PR though" what does "it" refer to?

rgrinberg commented:
I mean that:

* The PR in its current form looks ready to me.
* The higher level templates you mentioned can always be added later.

Sorry for the confusion. It seems like my vision for this command was quite a bit different than yours.

shonfeder commented:
@rgrinberg All clear! I agree with you and I plan to pursue PRs for the higher level templates in in subsequent work, just as you suggest.

No worries! Your vision *was* different than what I had in mind initially, but I was convinced by your reasoning along the way and I think we have ended up with something much more flexible that makes it trivial to achieve my initial aim, and will be useful for many other purposes besides. imo, it is fitting that someone in your position would push someone in mine to expand the depth and scope of their vision :)

shonfeder commented:
@diml That is a good point to consider. However, it seems that's more a problem of documentation than of the number of top-level commands. E.g., `git` has a [huge heap of top level commands](https://git-scm.com/docs/git), but only exposes about 20 of them via `--help`. While the current functionality here is lower-level than a quick start command, it is arguably much less low-level than `compute`, `printenv`, `rules`, `installed-libraries`, and `subst`. I can easily imagine `add` being used daily (just not by newbies).

diml commented:
> @diml That is a good point to consider. However, it seems that's more a problem of documentation than of the number of top-level commands. E.g., git has a huge heap of top level commands, but only exposes about 20 of them via --help.

That's true.

shonfeder commented:
If I understand correctly, the only outstanding concern here is with the naming? If I can get confirmation that the proposal to use `add` is rejected in favor of staying with `init`, I'll happily change that out, and then maybe this is ready for merging?

shonfeder commented:
@rgrinberg I believe all of your requested changes have been addressed (the qeustion of the name not withstanding). Is this correct?

rgrinberg commented:
@shonfeder It looked good to me, but awaiting approval from another maintainer. I will give it another read (and possible push some changes). Would you mind signing your commits meanwhile? It would not be possible to merge this work without that.

https://github.com/ocaml/dune/blob/master/CONTRIBUTING.md

rgrinberg commented:
Oh, and I believe there is consensus that `$ dune init` is the right name.

rgrinberg commented on bin/add.ml:
```diff
@@ -0,0 +1,120 @@
+  let report_invalid_option = function
+    | _, false -> ()  (* The option wasn't supplied *)
+    | option_name, true ->
+      Exn.fatalf
```
This should be `die`. That will give a nicer looking error message

rgrinberg commented on bin/add.ml:
```diff
@@ -0,0 +1,120 @@
+let info = Term.info "add" ~doc ~man
+
+let term =
+  let%map common_term = Common.term
```
Oh, and this needs to be converted to the 4.08 syntax.

rgrinberg commented on src/dune_add.ml:
```diff
@@ -0,0 +1,375 @@
+        | None -> Dune {f with content = f.content @ stanzas}
+        | Some (a, b) ->
+          Exn.fatalf
+            "Updating existing stanzas is not yet supported.\n\
```
This should also be die.

rgrinberg commented:
The code looks fine to me. @emillon could you give this a look next week? Just a high level review of the functionality is enough, the code is well isolated and I've already reviewed and it's fine.

@shonfeder once this is signed, it should be ready to go.

We can also discuss on how should we update the manual for this feature. It should be mentioned in the usage category at least.

shonfeder commented:
@rgrinberg Command name reverted to `dune init`, changes addressed, and commits signed off on. Feel free to push changes in. Naturally, I'm also happy to continue making reasonable changes necessary to move this towards closure.

After I making another round of changes needed to sync with upstream, some seemingly unrelated tests are failing:

```
$ make test
./_boot/default/bin/main_dune.exe runtest
Done: 5060/5068 (jobs: 3)File "test/blackbox-tests/test-cases/github717-odoc-index/run.t", line 1, characters 0-0:
        diff (internal) (exit 1)
(cd _build/default && /usr/bin/diff -u test/blackbox-tests/test-cases/github717-odoc-index/run.t test/blackbox-tests/test-cases/github717-odoc-index/run.t.corrected)
--- test/blackbox-tests/test-cases/github717-odoc-index/run.t	2019-03-09 20:48:20.000000000 -0500
+++ test/blackbox-tests/test-cases/github717-odoc-index/run.t.corrected	2019-03-09 20:55:00.000000000 -0500
@@ -1 +1,4 @@
   $ dune build @doc
+          odoc _doc/_odoc/pkg/hello_world/page-index.odoc
+  File "../../../../index.mld", line 0, characters 0-14:
+  '0': bad section level (2-4 allowed)
Done: 5060/5068 (jobs: 2)File "test/blackbox-tests/test-cases/odoc-unique-mlds/run.t", line 1, characters 0-0:
        diff (internal) (exit 1)
(cd _build/default && /usr/bin/diff -u test/blackbox-tests/test-cases/odoc-unique-mlds/run.t test/blackbox-tests/test-cases/odoc-unique-mlds/run.t.corrected)
--- test/blackbox-tests/test-cases/odoc-unique-mlds/run.t	2019-03-09 20:48:20.000000000 -0500
+++ test/blackbox-tests/test-cases/odoc-unique-mlds/run.t.corrected	2019-03-09 20:55:01.000000000 -0500
@@ -8,6 +8,12 @@
           odoc _doc/_odoc/lib/root.lib2/root_lib2.odoc
           odoc _doc/_html/root/Root_lib2/.dune-keep,_doc/_html/root/Root_lib2/index.html
           odoc _doc/_odoc/pkg/root/page-index.odoc
+  File "../../../_mlds/root/index.mld", line 0, characters 0-14:
+  '0': bad section level (2-4 allowed)
+  File "../../../_mlds/root/index.mld", line 4, characters 0-21:
+  only one title-level heading is allowed
+  File "../../../_mlds/root/index.mld", line 4, characters 0-21:
+  '1': bad section level (2-4 allowed)
           odoc _doc/_html/root/index.html
           odoc _doc/_html/root/Root_lib1/.dune-keep,_doc/_html/root/Root_lib1/index.html

@@ -20,9 +26,13 @@
           odoc _doc/_odoc/lib/scope1/scope1.odoc
           odoc _doc/_html/scope1/Scope1/.dune-keep,_doc/_html/scope1/Scope1/index.html
           odoc _doc/_odoc/pkg/scope1/page-index.odoc
+  File "../../../_mlds/scope1/index.mld", line 0, characters 0-16:
+  '0': bad section level (2-4 allowed)
           odoc _doc/_html/scope1/index.html
         ocamlc scope2/.scope2.objs/byte/scope2.{cmi,cmo,cmt}
           odoc _doc/_odoc/lib/scope2/scope2.odoc
           odoc _doc/_html/scope2/Scope2/.dune-keep,_doc/_html/scope2/Scope2/index.html
           odoc _doc/_odoc/pkg/scope2/page-index.odoc
+  File "../../../_mlds/scope2/index.mld", line 0, characters 0-16:
+  '0': bad section level (2-4 allowed)
           odoc _doc/_html/scope2/index.html
Done: 5060/5068 (jobs: 1)File "test/blackbox-tests/test-cases/odoc/run.t", line 1, characters 0-0:
        diff (internal) (exit 1)
(cd _build/default && /usr/bin/diff -u test/blackbox-tests/test-cases/odoc/run.t test/blackbox-tests/test-cases/odoc/run.t.corrected)
--- test/blackbox-tests/test-cases/odoc/run.t	2019-03-09 20:48:20.000000000 -0500
+++ test/blackbox-tests/test-cases/odoc/run.t.corrected	2019-03-09 20:55:01.000000000 -0500
@@ -4,6 +4,8 @@
           odoc _doc/_odoc/lib/bar/bar.odoc
           odoc _doc/_html/bar/Bar/.dune-keep,_doc/_html/bar/Bar/index.html
           odoc _doc/_odoc/pkg/bar/page-index.odoc
+  File "../../../_mlds/bar/index.mld", line 0, characters 0-13:
+  '0': bad section level (2-4 allowed)
           odoc _doc/_html/bar/index.html
           odoc _doc/_html/highlight.pack.js,_doc/_html/odoc.css
       ocamldep .foo_byte.objs/foo_byte.ml.d
@@ -17,6 +19,12 @@
           odoc _doc/_odoc/lib/foo/foo2.odoc
           odoc _doc/_html/foo/Foo2/.dune-keep,_doc/_html/foo/Foo2/index.html
           odoc _doc/_odoc/pkg/foo/page-index.odoc
+  File "../../../_mlds/foo/index.mld", line 0, characters 0-13:
+  '0': bad section level (2-4 allowed)
+  File "../../../_mlds/foo/index.mld", line 4, characters 0-20:
+  only one title-level heading is allowed
+  File "../../../_mlds/foo/index.mld", line 4, characters 0-20:
+  '1': bad section level (2-4 allowed)
           odoc _doc/_html/foo/index.html
           odoc _doc/_html/foo/Foo_byte/.dune-keep,_doc/_html/foo/Foo_byte/index.html
           odoc _doc/_html/foo/Foo/.dune-keep,_doc/_html/foo/Foo/index.html
make: *** [test] Error 1
```

Any advice on how to resolve these?

rgrinberg commented:
If you rebase on master, those should pass. The latest odoc was failing the tests, but I've fixed this issue in master already

shonfeder commented:
@rgrinberg This is rebased on the latest master already, and includes 14b570beee1adedb1488390081f61ab54cdefc8 . I tried a clean, build, test, but the test failures persist.

rgrinberg commented:
I see. It seems like it's just some shoddy bash scripting by me. One second.

shonfeder commented:
Thanks for looking in to the test failures, @rgrinberg! Let me know if I can
lend a hand.

I added a section on `dune init` to the usage, as suggested. It's just a
proposal to work with, and I'm happy to make any changes that may be desired
(and fine with leaving the documentation to someone else, if that's preferred).

rgrinberg commented:
The error in CI is now visibile:

```
File "test/blackbox-tests/test-cases/dune-init/run.t", line 1, characters 0-0:
        diff (internal) (exit 1)
(cd _build/default && /usr/bin/diff -u test/blackbox-tests/test-cases/dune-init/run.t test/blackbox-tests/test-cases/dune-init/run.t.corrected)
--- test/blackbox-tests/test-cases/dune-init/run.t	2019-03-10 04:20:39.303475305 +0000
+++ test/blackbox-tests/test-cases/dune-init/run.t.corrected	2019-03-10 04:21:11.784063358 +0000
@@ -227,8 +227,8 @@
   Hint: library names must be non-empty and composed only of the following characters: 'A'..'Z',  'a'..'z', '_'  or '0'..'9'
   [1]
   $ ls ./_test_lib
-  ls: ./_test_lib: No such file or directory
-  [1]
+  ls: cannot access ./_test_lib: No such file or directory
+  [2]
```

shonfeder commented on src/dune_init.ml:
```diff
@@ -20,7 +20,8 @@ module Kind = struct
     [ "exe", Executable
     ; "lib", Library
-    ; "test", Test ]
+    ; "test", Test
```
Thanks for all these fixes to bring the style in line with the idioms of the
project! I'll try to remember them for future contributions :)

If these kinds of formatting changes are significant enough to warrant
correction, I wonder if it would be worth codifying them under
https://github.com/ocaml/dune/blob/master/CONTRIBUTING.md#coding-style so you
can point contributors there.

I'll open a separate PR trying to put the principles behind these changes into
words there. That will help reinforce them for my own sake as well.

rgrinberg replied:
Soon we'll have ocamlformat forcing its style on us. So there's really no necessity.

shonfeder replied:
Ah, nice!

shonfeder commented on src/dune_init.ml:
```diff
@@ -20,7 +20,8 @@ module Kind = struct
     [ "exe", Executable
     ; "lib", Library
-    ; "test", Test ]
+    ; "test", Test
```
This commit is missing the sign off.

rgrinberg replied:
Yes indeed. I was planning to squash into your commit anyways.

rgrinberg commented on src/dune_lang/dune_lang.mli:
```diff
@@ -146,9 +148,11 @@ module Cst : sig
 
   val concrete : Ast.t -> t
 
+  val to_sexp : t -> sexp option
```
I know this is just following the convention in this module, but `sexp` is actually the wrong name. `Sexp.t` is just a debugging format that we use. Here it refers to dune's sexp based language which is actually `Dune_lang.t`.

Earlier, the distinction between the two wasn't so clear so things were messy. We have to think of a new name here and for other misuses of `sexp`.

shonfeder replied:
Maybe something along the lines of `pure_ast`?

rgrinberg replied:
Unfortunately, the `ast` is also taken. It's just the dune language + location information.

shonfeder replied:
Yes, that's why I suggested the `pure` prefix.

shonfeder commented on src/dune_init.ml:
```diff
@@ -126,8 +124,8 @@ module File = struct
                A preexisting dune stanza conflicts with a generated stanza:\n\n\
-               Generated stanza:\n%s\n\nPre-existing stanza:\n%s"
-            (to_string a) (to_string b)
+               Generated stanza:\n%a\n\nPre-existing stanza:\nas"
```
Shouldn't the last conversion read `%a`?

rgrinberg replied:
Yes, that's a good point. It's quite annoying that `die` swallows things this way because its return type is polymorphic

rgrinberg commented:
@shonfeder have a look at my last commit. I believe this root argument isn't necessary.

rgrinberg commented:
@shonfeder I'm curious, why didn't you just use dune's own parsing code for parsing the stanzas? This way, I think we have some duplication in the code and it isn't very desirable.

shonfeder commented:
@rgrinberg Can you please specify which part of the code you think introduces parsing duplication and what existing code you think I might use instead? Afaik, I do very little in the way of parsing here. The only exception that comes to mind is the temporary function https://github.com/ocaml/dune/pull/1448/files#diff-ec22fb5fe4ab585936d484f498556279R69, which has the very specific purpose of identifying conflicting stanzas from the already loaded `Cst.t`.

shonfeder commented:
Also, note that we need to preserve comments, since we read in and then re-emit dune code in dune files that may be hand written.

shonfeder commented on src/dune_init.ml:
```diff
@@ -172,29 +172,23 @@ end
       match path with
-      | None -> root
-      | Some path ->
-        if Filename.is_relative path then
```
Without this, are users still able to specify either an absolute or a relative path as the location for the added component?

shonfeder replied:
I think I forgot to include a test case for that behavior. I'll try to add that tonight if I get time.

rgrinberg replied:
Path.of_string should be smart enough to interpret relative paths correctly. A test would be useful indeed.

rgrinberg replied:
Path.of_string should be smart enough to interpret relative paths correctly. A test would be useful indeed.

rgrinberg commented:
@shonfeder yeah, that's indeed the function I'm talking about. Why is there an issue here regarding comments btw? We are simply using the `Cst.t` to create `Signature.t` values.

My idea would be to get rid of `Signature.t` altogether and just use `Dune_file.Stanzas.parse` along with the `Stanza.t` type to detect exes/libs. Do you think that's reasonable?

Also note that we if we initialize a "super context", then we'll have all the parsed stanzas for free.

shonfeder commented:
> My idea would be to get rid of Signature.t altogether and just use Dune_file.Stanzas.parse along with the Stanza.t type to detect exes/libs. Do you think that's reasonable?

Yep. That's a better idea for identifying conflicts. I'll make the change this weekend.

> Why is there an issue here regarding comments btw?

The issue with the comments in general is this: since we aim to be able to extend existing dune files and stanzas, we have to preserve the comments which users may have added. This means that in general, we can't *just* work with the `Stanza.t` type, or else we'd lose the comments when we write the stanzas back to the file. We can use `Stanza.t` for conflict detection, but we cannot simply use `Stanza.t` once we move to inserting dependencies, because we'll end up erasing user comments.

> Also note that we if we initialize a "super context", then we'll have all the parsed stanzas for free.

I wasn't aware of `super_context`. That's quite cool. I think this would face the same problem of wiping out user comments tho, no?

rgrinberg commented:
> I wasn't aware of super_context. That's quite cool. I think this would face the same problem of wiping out user comments tho, no?

Indeed, but it it can be used for just identify conflicts. The dune file can then be re-parsed to cst's when necessary.

shonfeder commented:
@rgrinberg In order for the planned code reuse to work in a sensible way, the new stanza (of type `Cst.t`) must also be converted to a `Stanza.t`. So far, I have not been able to find any existing function `Cst.t -> Stanza.t` (nor any function to provide a helpful segment of that transformation, such as `Ast.t -> Stanza.t`). Do you happen to know of a sensible way to make this conversation?  

I would, of course, be happy to work out this plumbing myself, but if that is required, I think we should just mark this with a `TODO` and address it in a follow up PR. This PR has already suffered from enough gradual sprawl, feature creep, and friction from staleness, and it would really help with my momentum to land it without too much more ado.

rgrinberg commented:
Yeah, that's fine with me as well. Let's do the following:

* Makes sure everything is signed-off/rebased on master
* Update the manual and add a little note that `$ dune init` is still in development and is subject to change without warning
* Merge this PR.

kevinji commented on test/blackbox-tests/test-cases/dune-init/run.t:
```diff
@@ -0,0 +1,281 @@
+
+  $ rm -rf ./_test_tests_dir
+
+Adding componets to default and non-standard places
```
s/componets/components/

kevinji commented on test/blackbox-tests/test-cases/dune-init/run.t:
```diff
@@ -0,0 +1,281 @@
+
+  $ rm -rf ./_test_dir
+
+Add a library to a dune file in a directory specififed with an absolute path
```
s/specififed/specified/

kevinji reviewed (commented):
Found a few typos.

shonfeder commented:
Thanks for catching those, @kevinji -- fixed now.

@rgrinberg action items addressed. Unless something else comes up, I think this should be ready for merge once CI is finished.

shonfeder commented:
After discussion on #1972, I've learned I was misunderstanding how `Dune_file.Stanzas.parse` worked. Witb my understanding corrected, I think I can now eliminate the ad-hoc parsing this PR.

I also need to correct a documentation comment is added to that function in this PR.

rgrinberg commented:
@shonfeder Please make this improvement in a subsequent PR. Thank you for all the hard work on this. Let's track future improvements in other tickets.
